### PR TITLE
fix concurrency bugs

### PR DESCRIPTION
**Pull request by pwinckles**
# What does this Pull Request do?
Fixes a few concurrency bugs in Fedora's ocfl code.

# Interested parties
@fcrepo4/committers

### CODE REVIEW

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java [hunk 1]
```diff
@@ -61,23 +61,20 @@ public PersistentStorageSession getSession(final String sessionId) {
...
                 fedoraOcflIndex,
-                objectSessionFactory);
-        sessionMap.put(sessionId, newSession);
-        return newSession;
+                objectSessionFactory));
     }
 
     @Override
     public PersistentStorageSession getReadOnlySession() {
         if (this.readOnlySession == null) {
```
Comment (whikloj) [inline fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java:72]: Is this line needed as it is repeated inside the `synchronized (this) {` block? I feel like it was just an oversight?

Comment (pwinckles) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java:72]: No synchronization is required.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent(K,java.util.function.Function)

>  The entire method invocation is performed atomically, so the function is applied at most once per key. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this map.

Comment (pwinckles) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java:72]: Whoops, I got confused about which line you were referring to and thought you meant L64.

Comment (pwinckles) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java:72]: Do you mean L72?

If so, then yes, both conditionals are needed. https://en.wikipedia.org/wiki/Double-checked_locking

Comment (whikloj) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java:72]: Ah okay, thanks for the explanation.

**Pull request merged**